scapy/layers/ipsec.py

```diff
@@ -115,6 +131,17 @@ class ESP(Packet):
         XStrField('data', None),
     ]
 
+    @classmethod
+    def dispatch_hook(cls, _pkt=None, *args, **kargs):
+        if _pkt:
+            if len(_pkt) >= 4 and struct.unpack("!I", _pkt[0:4])[0] == 0x00:
+                return NON_ESP
+            elif len(_pkt) == 1 and struct.unpack("!B", _pkt)[0] == 0xff:
+                return NAT_KEEPALIVE
+            else:
+                return ESP
+        return cls
```

Comment:
👍 It's actually much nicer and more natural to have the dispatch hook here in ESP instead of an artificial UDP_ENCAP class.

https://github.com/secdev/scapy/blob/43582d36fe202916fab4f36f0fd10d697eabd8a5/scapy/contrib/ikev2.py#L912-L927